src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -815,117 +818,105 @@ private async ValueTask<Memory<byte>> StartWriteAsync(int writeBytes, Cancellati
-                    await new ValueTask(FlushOutgoingBytesAsync()).ConfigureAwait(false); // await ValueTask to minimize number of awaiter fields
+                    int activeBufferLength = _outgoingBuffer.ActiveLength;
+                    if (writeBytes >= totalBufferLength - activeBufferLength && activeBufferLength > 0)
+                    {
+                        // We explicitly do not pass cancellationToken here, as this flush impacts more than just this operation.
+                        await new ValueTask(FlushOutgoingBytesAsync()).ConfigureAwait(false); // await ValueTask to minimize number of awaiter fields
+                    }
                 }
 
+                // Invoke the callback with the supplied state and the target write buffer.
                 _outgoingBuffer.EnsureAvailableSpace(writeBytes);
-                Memory<byte> writeBuffer = _outgoingBuffer.AvailableMemory.Slice(0, writeBytes);
-                _currentWriteSize = writeBytes;
-
-                return writeBuffer;
+                lockedAction(state, _outgoingBuffer.AvailableMemorySliced(writeBytes));
             }
-            catch
+            finally
             {
                 _writerLock.Exit();
-                throw;
             }
         }
 
         /// <summary>Flushes buffered bytes to the wire.</summary>
+        /// <param name="bytesWritten">The number of bytes written that are being flushed.</param>
         /// <param name="flush">When a flush should be performed for this write.</param>
         /// <remarks>
         /// Writes here need to be atomic, so as to avoid killing the whole connection.
-        /// Callers must hold the write lock, which this will release.
+        /// Callers must hold the write lock.
         /// </remarks>
-        private void FinishWrite(FlushTiming flush)
+        private void FinishWrite(int bytesWritten, FlushTiming flush)
         {
-            if (NetEventSource.IsEnabled) Trace($"{nameof(flush)}={flush}");
-
             // We can't validate that we hold the mutex, but we can at least validate that someone is holding it.
             Debug.Assert(_writerLock.IsHeld);
 
-            _outgoingBuffer.Commit(_currentWriteSize);
-            _lastPendingWriterShouldFlush |= (flush == FlushTiming.AfterPendingWrites);
-            EndWrite(forceFlush: (flush == FlushTiming.Now));
-        }
-
-        private void CancelWrite()
-        {
-            if (NetEventSource.IsEnabled) Trace("");
-
-            // We can't validate that we hold the mutex, but we can at least validate that someone is holding it.
-            Debug.Assert(_writerLock.IsHeld);
+            if (NetEventSource.IsEnabled) Trace($"{nameof(flush)}={flush}");
 
-            EndWrite(forceFlush: false);
+            _outgoingBuffer.Commit(bytesWritten);
+            _lastPendingWriterShouldFlush |= flush == FlushTiming.AfterPendingWrites;
```

Comment:
Almost certainly inconsequential, but do CPUs optimize out this load/store if the value being ORed is 0?